### PR TITLE
Feature: Allow encrypting sensitive fields in export

### PR DESCRIPTION
**Pull request by stumpylog**
## Proposed change

Allows a user to specify a passphrsae when exporting, and the mail account password will be stored encrypted.  When the same passphrase is provided on import, the password is decrypted again.

If the passphrase is not provided on import, the import will fail.

The encryption is based on the provided link.  The necessary parameters to reconstruct the key are stored in the newly renamed `metadata.json` which replaces `version.json` with a more useful name and potential for adding more data in the future.  The values stored here are safe to be public.  Django, for example, stores these in the field for passwords, so it's public already.

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
This is cool stumpy, thanks for working on it. Not sure if youre still working on it but I think a docs update is missing (usually the least exciting part of a PR :)

**Comment by shamoon** [comment]
Lol, never mind!

File: src/documents/tests/test_management_exporter.py [hunk 1]
```diff
@@ -840,3 +842,139 @@ def test_export_data_only(self):
...
+            "document_exporter",
+            "--no-progress-bar",
+            str(self.target),
+            stdout=stdout,
+        )
+        stdout.seek(0)
+        self.assertIn(
+            (
+                "You have configured mail accounts, "
+                "but have no passphrase was given. "
```
Comment (shamoon) [inline src/documents/tests/test_management_exporter.py:976]:
```suggestion
                "but no passphrase was given. "
```

File: src/documents/management/commands/document_exporter.py [hunk 2]
```diff
@@ -527,3 +544,23 @@ def check_and_copy(
...
+
+            for mail_account_record in manifest["mail_accounts"]:
+                mail_account_record["fields"]["password"] = self.encrypt_string(
+                    value=mail_account_record["fields"]["password"],
+                )
+        elif MailAccount.objects.count() > 0:
+            self.stdout.write(
+                self.style.NOTICE(
+                    "You have configured mail accounts, "
+                    "but have no passphrase was given. "
```
Comment (shamoon) [inline src/documents/management/commands/document_exporter.py:563]:
```suggestion
                    "but no passphrase was given. "
```

File: src/documents/management/commands/document_exporter.py [hunk 3]
```diff
@@ -165,6 +166,11 @@ def add_arguments(self, parser):
             help="If set, the progress bar will not be shown",
         )
 
+        parser.add_argument(
+            "--passphrase",
+            help="If provided, is used to encrypt mail account passwords in the export",
```
Comment (shamoon) [inline src/documents/management/commands/document_exporter.py:171]:
```suggestion
            help="If provided, is used to encrypt sensitive data in the export",
```

maybe just making this more generic for the future?

**Review by shamoon — APPROVED** [review]
Awesome work stumpy, I dont have much to say code-review wise (other than the unimportant notes) but works just as expected in my testing. Also seems really solid footing to be expanded upon.

Screenshot 2024-06-06 at 1 02 39 PM

**Comment by tribut** [comment]
Is there a reason we aren't encrypting the whole export when a passphrase is given? It would be much safer in terms of other (future) secrets and possibly easier to handle as well.

**Comment by stumpylog** [comment]
Yes, it would be a huge file size increase

**Pull request merged**